rng-tools: 6.7 -> 6.8

c0bw3b opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Update
\+ enable tests
\+ enable jitterentropy by default
\+ add c0bw3b to maintainers

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @JohnAZoidberg

c0bw3b committed:
rng-tools: 6.7 -> 6.8

+ run tests
+ enable jitterentropy by default
+ add c0bw3b to maintainers

c0bw3b commented:
Changelog:

> Bug fix update containing the following fixes:
> 
> + Fixed texrels on on rdrand_asm.S for pic compilation
> + allow use of libargp if libc lacks argp parsing
> + explicitly link against -lcrypto, fixing build in pkcs11 entropy source
> + replace pthread_yield with posix compliant sched_yield
> + bias rngd to use faster sources of entropy when available, falling back to slower sources when needed
> + Fix a shutdown delay resulting from a thread exit race
> + Fix a few minor compilation warnings
> + Fix make distcheck make target
> + Minor typo fixes/cleanups

c0bw3b commented:
@GrahamcOfBorg eval

JohnAZoidberg commented:
Thanks! I looked at the individual patches and noticed that argp-standalone can now be used (https://github.com/nhorman/rng-tools/commit/ddecdb58b032b77f4519fccdb1b57df9eebb5ce0). It's needed when using musl because that doesn't include argp.

`argp-standalone` can just be added to `buildInputs`. The configure script only links it, if the libc doesn't provide it. I checked that the closure doesn't include it, when building with glibc.

I don't know, however, how to build just this package with musl to verify that that works.
Do you?

`nix-build -A pkgsCross.musl64.rng-tools` and `nix-build -A pkgsMusl.rng-tools` seem to build all dependencies with musl, too.

c0bw3b commented:
> I don't know, however, how to build just this package with musl to verify that that works.
> Do you?

Not exactly. I would have tried `build pkgsMusl.rng-tools` like you did.
But @dtzWill could help us here. :)

c0bw3b commented:
I added `argp-standalone`. It was also mentioned in the changelog in fact.
I also removed the file copy in postPatch because it doesn't seem to serve any purpose. (?) The file is not installed.

kmcopper commented:
Would it be possible to change enableFeature withJitterEntropy to withFeature as well as enable it by default? This is the default in two other distributions, archlinux and fedora, and shouldn't harm entropy at all, only help it. Otherwise looks good. I tested this on my own branch (https://github.com/c0bw3b/nixpkgs/compare/pkg/rngtools...kmcopper:rng-tools) and it works great.

Closure size (with my branch):
rng-tools-6.7 45981904 -> 46038872
Tested, compiled, and ran on NixOS in a VM

c0bw3b commented:
Since this v6.8 release `rngd` will prioritize faster hardware sources (hwrng, TPM, RDRAND) over software sources like jitterentropy, which may still be used as additional source if available. Plus the `jitterentropy` lib has shown [compliance with NIST SP800-90B](http://www.chronox.de/jent.html) as an example benchmark.
For me the concern about its quality is gone by now and I'm +1 on enabling it by default along other sources. The additional closure size is minimal (see further down).
I'll update this PR soon.

> Would it be possible to change enableFeature withJitterEntropy to withFeature

Upstream configure flag to add jitter source is `--enable-jitterentropy`  and not `--with-jitterentropy` like other features (`--with-pkcs11` for example).
So it has to be `enableFeature` here.

> and shouldn't harm entropy at all, only help it. 

Entropy is more complicated than that. You can actually harm it with too many sources gathering too many data points from not-so-unpredictable events. See Dan J. Bernstein on this topic:
http://blog.cr.yp.to/20140205-entropy.html
Plus another good read on the matter of randomness and entropy in general : https://www.2uo.de/myths-about-urandom/

> Additionally this will allow rngd to improve entropy on sandybridge systems.

Sandy Bridge has RDRAND I believe? Plus it would help rngd only if you don't have any other sources available. On my test system with RDRAND and TPM sources available, the average "input channel speed" remains the same with or without jitter source.

---

Regarding closure size:
```console
# standard build without jitterentropy:
/nix/store/c00nxwcn3x3x1n87hidbgqgvln2xbwq8-rng-tools-6.8	  80.0K	  43.9M

# rng-tools build WITH jitterentropy:
/nix/store/xb347wd2bvqn1fh7p6db2b75lfdfm5i3-rng-tools-6.8	  97.3K	  43.9M
```

kmcopper commented:
My testing was done on a 2600k. RDRAND was added in Ivybridge (Sandybridge tock).

kmcopper reviewed (approved):
My testing was done on a 2600k. Looks good to me.
RDRAND was added in Ivybridge (Sandybridge tock).

c0bw3b commented:
@GrahamcOfBorg build rng-tools
